### ISSUE TITLE
Delay dispatching actions

### ISSUE THREAD

**Issue by ms88privat** [jfairbank/redux-saga-test-plan#69]
The integration testing part looks really good!

But is there the possibility to delay the dispatch of actions?

Here my current Example:

```js

    it('should limit save operations', () => {
        const saveResourceAction = createSaveResourceActionMock();
        const delayTime = 4000;
        const maxOperations = 2;

        return expectSaga(watchSaveResource, delayTime, maxOperations)
            .call(saveResourceAction.payload.resolve)
            .call(saveResourceAction.payload.resolve)
            .call(saveResourceAction.payload.reject, slowDownError)
            .call(saveResourceAction.payload.reject, slowDownError)
            .put({ type: 'FEEL_FREE' })
            .dispatch(saveResourceAction)
            .dispatch(saveResourceAction)
            .dispatch(saveResourceAction)
            .dispatch(saveResourceAction)
            .run(5001)
            .then(() => {
                expect(saveResourceAction.payload.resolve.mock.instances.length).toBe(2);
                expect(saveResourceAction.payload.reject.mock.instances.length).toBe(2);
            });
    }, 6001);

```

The logic is as follows:

- handle incoming Actions -> resolve them
- after "maxOperations" actions handled -> reject all other
- stop rejecting after delayTime ('FEEL_FREE' action gets put)
- repeat

Because the test run dispatches every action in no time, the saga will handle two of them and reject the other two. What I also would like to test is, that if I dispatch two actions, then wait for greater then delayTime and dispatch another two actions, these will get also handled instead of rejected.

It should look something like this maybe:

```js

    it('should limit save operations', () => {
        const saveResourceAction = createSaveResourceActionMock();
        const delayTime = 4000;
        const maxOperations = 2;

        return expectSaga(watchSaveResource, delayTime, maxOperations)
            .call(saveResourceAction.payload.resolve)
            .call(saveResourceAction.payload.resolve)
            .call(saveResourceAction.payload.resolve)
            .call(saveResourceAction.payload.resolve)
            .put({ type: 'FEEL_FREE' })
            .dispatch(saveResourceAction)
            .dispatch(saveResourceAction)
            .testHelper.delay(delayTime + 100)
            .dispatch(saveResourceAction)
            .dispatch(saveResourceAction)
            .run(5001)
            .then(() => {
                expect(saveResourceAction.payload.resolve.mock.instances.length).toBe(4);
                expect(saveResourceAction.payload.reject.mock.instances.length).toBe(0);
            });
    }, 6001);

```

Do you have any remarks on this? Thanks

**Comment by jfairbank** [comment]
Hi, @ms88privat!

I had thought about the scenario where one might want to delay dispatching actions and believe it's definitely a common use case. I'm not opposed to adding the API, but I think a good first step in general might be to allow actions to be dispatched after the saga starts running. That would initially allow sagas with delayed actions to be tested:

```js
import { put, take } from 'redux-saga/effects';
import { expectSaga } from 'redux-saga-test-plan';

const delay = time => new Promise(resolve => setTimeout(resolve, time));

function* mainSaga() {
  yield take('FOO');
  yield take('BAR');
  yield put({ type: 'DONE' });
}

it('can dispatch actions while running', () => {
  const saga = expectSaga(mainSaga);

  saga.put({ type: 'DONE' });

  saga.dispatch({ type: 'FOO' });

  const promise = saga.run({ timeout: false });

  return delay(500).then(() => {
    saga.dispatch({ type: 'BAR' });
    return promise;
  });
});
```

Granted, the above test is a little more cumbersome to write, so I'm still open to exploring additional API support for this after tweaking how actions are dispatched internally. What do you think?

**Comment by jfairbank** [comment]
@ms88privat delaying is available in v2.3.0 now

You can learn more [here](http://redux-saga-test-plan.jeremyfairbank.com/integration-testing/dispatching.html).
